Source/qol/chatlog.cpp

```diff
@@ -0,0 +1,209 @@
+constexpr int ContentTextWidth = 577;
+
+int LineHeight()
+{
+	return IsSmallFontTall() ? 18 : 14;
+}
+
+int BlankLineHeight()
+{
+	return 12;
+}
+
+int DividerLineMarginY()
+{
+	return BlankLineHeight() / 2;
+}
+
+int HeaderHeight()
+{
+	return PaddingTop + LineHeight() + 2 * BlankLineHeight() + DividerLineMarginY();
+}
+
+int ContentPaddingY()
+{
+	return BlankLineHeight();
+}
+
+int ContentsTextHeight()
+{
+	return PanelHeight - HeaderHeight() - DividerLineMarginY() - 2 * ContentPaddingY() - BlankLineHeight();
+}
+
+int NumVisibleLines()
+{
+	return (ContentsTextHeight() - 1) / LineHeight() + 1; // Ceil
+}
+
+} // namespace
+
+bool ChatLogFlag = false;
+
+void ToggleChatLog()
+{
+	if (ChatLogFlag) {
+		ChatLogFlag = false;
+	} else {
+		stextflag = STORE_NONE;
+		invflag = false;
+		chrflag = false;
+		sbookflag = false;
+		spselflag = false;
+		if (qtextflag && leveltype == DTYPE_TOWN) {
+			qtextflag = false;
+			stream_stop();
+		}
+		QuestLogIsOpen = false;
+		HelpFlag = false;
+		AutomapActive = false;
+		CancelCurrentDiabloMsg();
+		gamemenu_off();
+		SkipLines = 0;
+		ChatLogFlag = true;
+		doom_close();
+	}
+}
+
+void AddMessageToChatLog(const std::string &message, Player *player, UiFlags flags)
+{
+	MessageCounter++;
+	time_t tm = time(nullptr);
+	std::string timestamp = fmt::format("[#{:d}] {:02}:{:02}:{:02}", MessageCounter, localtime(&tm)->tm_hour, localtime(&tm)->tm_min, localtime(&tm)->tm_sec);
+	int oldSize = ChatLogLines.size();
+	ChatLogLines.emplace_back(MultiColoredText { "", { {} } });
+	if (player == nullptr) {
+		ChatLogLines.emplace_back(MultiColoredText { "{0} {1}", { { timestamp, UiFlags::ColorRed }, { message, flags } } });
+	} else {
+		std::string playerInfo = fmt::format(_("{:s} (lvl {:d}): {:s}"), player->_pName, player->_pLevel, "");
+		ChatLogLines.emplace_back(MultiColoredText { message, { {} }, 20 });
+		UiFlags nameColor = player == &Players[MyPlayerId] ? UiFlags::ColorWhitegold : UiFlags::ColorBlue;
+		ChatLogLines.emplace_back(MultiColoredText { "{0} - {1}", { { timestamp, UiFlags::ColorRed }, { playerInfo, nameColor } } });
+	}
+
+	int diff = ChatLogLines.size() - oldSize;
+	// only autoscroll when on top of the log
+	if (SkipLines != 0) {
+		SkipLines += diff;
+		UnreadFlag = true;
+	}
+}
+
+void DrawChatLog(const Surface &out)
+{
+	DrawSTextHelp();
+	DrawQTextBack(out);
+
+	if (SkipLines == 0) {
+		UnreadFlag = false;
+	}
+
+	const int lineHeight = LineHeight();
+	const int blankLineHeight = BlankLineHeight();
+	const int sx = PANEL_X + PaddingLeft;
+	const int sy = UI_OFFSET_Y;
+
+	DrawString(out, fmt::format(_("Chat History (Messages: {:d})"), MessageCounter),
+	    { { sx, sy + PaddingTop + blankLineHeight }, { ContentTextWidth, lineHeight } },
+	    (UnreadFlag ? UiFlags::ColorRed : UiFlags::ColorWhitegold) | UiFlags::AlignCenter);
+
+	char timestamp[64];
+	time_t tm = time(nullptr);
+	sprintf(timestamp, "%02d:%02d:%02d", localtime(&tm)->tm_hour, localtime(&tm)->tm_min, localtime(&tm)->tm_sec);

```

Comment:
Could we use a std::string and fmt for this?

Reply:
aye